### PR TITLE
Handle Java sandboxed caching better

### PR DESCRIPTION
**Pull request by DomT4**
We talked about this a while back, but for the benefit of new maintainers and/or curious contributors/users as part of the sandbox we've barred things from writing outside cache locations, the buildpath, testpath, etc during installation and tests.

Tim has an open PR to set this globally but at the moment there are an unknown number of formulae that break if you pass `_JAVA_OPTIONS` to them: https://github.com/Homebrew/homebrew/pull/44567. Consequently, we can't really set it globally until upstream move on it & we can do extensive testing in the core.

So far we've been setting some variation of `_JAVA_OPTIONS` in `def install` and `test do`, using one of the following as the variable:
* `ENV["HOME"]` - Pretty much means buildpath/testpath.
* `buildpath` or `testpath`
* `buildpath/".some_hidden_dir"`

The problem with all the above is that it kills off one of the useful features of cache directories in that once you've downloaded the required elements if the build fails you can re-do the build without having to download everything afresh. Some Java-using formulae are grabbing 100+ MB of files during installation, so we're taking liberties with user bandwidth at times under the present system.

The other significantly more minor problem is that using a bunch of different variable names leaves contributors confused as to which they should actually use, something we see pretty often. A single unified `java_cache` option is easier to communicate.

No direct sandbox changes are required because the current sandbox mechanism allows file/folder reading/writing in the cache directories during both install and test.

The `@private` tag in `formula.rb` is intended to mean "Don't use this outside of the core". I appreciate the downsides on that is that people may choose to ignore this. If necessary, rather than scrapping the option entirely once the time comes to make the `_JAVA_OPTIONS` variable global we could just do something like:

```ruby
old_java_options = ENV["_JAVA_OPTIONS"]
ENV.append "_JAVA_OPTIONS", "-Duser.home=#{java_cache}"
```
Which means we won't have to change in-formula code immediately *or* have the code break for people using the private DSL externally. If people prefer that over the shorthand simply being killed at a future date, I'll adjust the notations in `formula.rb` accordingly.

Let the brutal criticism commence! :wink:

### CODE REVIEW

**Comment by dunn** [comment]
Oh, cool.  I'd been thinking about how to cache all the stuff that Maven and Gradle builds download; maybe eventually we should try to have something like 
```ruby
depends_on "maven"
def install
  mvn "wherever"
````
which would run `mvn package` with the home dir and whatever else set in order to prevent downloads during install (after the first).

**Comment by mikemcquaid** [comment]
> `mvn "wherever"`

This approach would work for me; we've done similar things with `xcodebuild` and `ruby`.

**Comment by mikemcquaid** [comment]
Works for me. I wonder if we're gonna do this if it would be nicer still to have a `ENV.java_cache` (and/or the `mvn` @dunn suggested) which sets this up for you automagically.

**Comment by DomT4** [comment]
> I wonder if we're gonna do this if it would be nicer still to have a ENV.java_cache

I like this option. Will play with that locally and make some changes.

The problem goes beyond maven usage, a broader solution is going to apply better. Java routinely attempts to violate the sandbox, the saving grace is that most of that is non-fatal to the build.

Agree long-term ideally all formulae should download all necessary objects at the start of the build and then incoming/outgoing network connections should be barred.

**Comment by mikemcquaid** [comment]
> I like this option. Will play with that locally and make some changes.

:cool:. The other benefit with that is we can change how it works under the hood without needing to tweak formulae code.

**Comment by DomT4** [comment]
Alright, this is ready for another review.

File: Library/Homebrew/extend/ENV/std.rb [hunk 1]
```diff
@@ -303,6 +303,10 @@ def libstdcxx
     end
   end
 
+  def java_cache
+    append "_JAVA_OPTIONS", "-Duser.home=#{HOMEBREW_CACHE}/java_cache"
+  end
```
Comment (bfontaine) [inline Library/Homebrew/extend/ENV/std.rb]: Shouldn’t we put this in `ENV/shared.rb` since it’s included in both `std` and `super`?

Comment (DomT4) [inline Library/Homebrew/extend/ENV/std.rb]: Doing so would seem logical, but this style mirrors the choices made for `deparallelize`. I'm not too fussed where it goes. If people are happy with shared, `¯\_(ツ)_/¯`

File: Library/Homebrew/cmd/cleanup.rb [hunk 2]
```diff
@@ -62,6 +62,10 @@ def cleanup_cache
         cleanup_path(path) { path.unlink }
         next
       end
+      if path.to_s.include?("java_cache")
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/cleanup.rb]: I don't really like the idea of allowing matches at arbitrary positions inside the string, though it's quite unlikely earlier path components will match in practice. Since we're explicitly iterating over children of the `HOMEBREW_CACHE` directory, why not use something like `if path.basename.to_s == "java_cache"`, possibly even with appended `&& path.directory?` as we know this is supposed to be a directory?

Comment (DomT4) [inline Library/Homebrew/cmd/cleanup.rb]: This style was mimicked from below (I have a thing for consistency, heh). I've no strong objection to checking the path is actually a path, but I think the chances of stumbling across a file including `java_cache` that resides inside our `HOMEBREW_CACHE` is pretty low.

File: Library/Homebrew/extend/ENV/std.rb (hunk 1 above)
Comment (UniqMartin) [inline Library/Homebrew/extend/ENV/std.rb]: Only there are subtle differences in the implementation of `deparallelize` in `std` and `super`, while this method is identical and thus shouldn't be duplicated. (Not saying that those differences in `deparallelize` should exist. But moving/unifying `deparallelize` is not what this PR is about.)

Comment (DomT4) [inline Library/Homebrew/extend/ENV/std.rb]: Heh, so there are. I can't read for crap, apparently. Made the change locally after Baptiste's comment anyhow, just hadn't pushed/tested yet so doing that at the moment.

File: Library/Homebrew/cmd/cleanup.rb (hunk 2 above)
Comment (UniqMartin) [inline Library/Homebrew/cmd/cleanup.rb]: The code of the `clean` command is probably not the best example to learn from. :wink: I'm very much in favor of doing the right and safe thing, especially if it is no big deal, instead of wiping unrelated stuff that happens to have the “wrong” file name (even if the effect is limited to the cache directory).

Comment (DomT4) [inline Library/Homebrew/cmd/cleanup.rb]: The code directly below that this mimics was added and reviewed very recently :wink:. Honestly, if someone can find me an example in the wild of a formula we download that happens to use `java_cache` in the file or folder name somewhere I'll send them the standard offering of :cake:. I'm perfectly happy to make the cache name a whole lot more random if desired.

File: Library/Homebrew/extend/ENV/std.rb (hunk 1 above)
Comment (UniqMartin) [inline Library/Homebrew/extend/ENV/std.rb]: No worries! We all tend to read what we want to read instead of what is actually written. :smirk_cat: Just wanted to point out the difference, even if you've already moved the stuff.

File: Library/Homebrew/cmd/cleanup.rb (hunk 2 above)
Comment (UniqMartin) [inline Library/Homebrew/cmd/cleanup.rb]: Simple example: If `HOMEBREW_CACHE` happens to be `/Users/java_cache/homebrew/cache` (I know that `java_cache` is a pretty stupid user name) then this will wipe out the entire cache every time `brew cleanup` is invoked.

> I'm perfectly happy to make the cache name a whole lot more random if desired.

Please let's not go down that route. I don't see why we shouldn't use a 100% safe and simple alternative instead of leaving a nonzero (if small) chance of things going completely wrong.

Comment (bfontaine) [inline Library/Homebrew/cmd/cleanup.rb]: +1 for the 100% safe directory name as well. Can we still have some :cake:? :smile:

**Comment by dunn** [comment]
Would it be safe to automatically set it when `depends_on :java`? https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/requirements/java_requirement.rb#L15

**Comment by DomT4** [comment]
> Would it be safe to automatically set it when depends_on :java?

Alas, no. See the issue referenced in the initial (now outdated) explanation.

**Comment by dunn** [comment]
:see_no_evil:

File: Library/Homebrew/cmd/cleanup.rb (hunk 2 above)
Comment (DomT4) [inline Library/Homebrew/cmd/cleanup.rb]: You could make the same statement about any cache directory/parent which includes `--` in the name in that case, which seems *considerably* more likely. I'm not sure we should necessarily be devoting an enormous amount of time to protecting people against Homebrew-related Darwinism, of which there are inevitably a bunch of tripwires for in the core.

>  I don't see why we shouldn't use a 100% safe and simple alternative instead of leaving a nonzero (if small) chance of things going completely wrong.

I think it's reasonable to try and strike a balance between protecting everyone under the sun using Homebrew from self-inflicted "Doh" moments and speed.

A rough average across 5 benchmarks of each shows the suggested method is 0.5 to 0.9 seconds slower per run. We've been fussy about less than that elsewhere, many many many many times :sweat:

Comment (DomT4) [inline Library/Homebrew/cmd/cleanup.rb]: I'll change if if people are that in love with it, but I kind of feel like if we start trying to protect people from every funky edge case "Doh" moment in the core possible we're all going to spend the next 60 years making some very minor core tweaks.

**Comment by DomT4** [comment]
@dunn The dream is that one day we can set this universally in `formula.rb` and have it completely invisible to users, but we're a way from that yet :cry:

File: Library/Homebrew/cmd/cleanup.rb (hunk 2 above)
Comment (UniqMartin) [inline Library/Homebrew/cmd/cleanup.rb]: I have nothing to add, I'm afraid. :disappointed:
